DemuxFastqs improvements

nh13 opened the pull request:
Three main features have been added:
1. The output metrics file name is no longer required.
2. A CSV file can be given with sample metadata in lieu of a sample sheet.  Some users requested this feature as they are not experts on Illumina sample sheets.
3. Gzipped FASTQs can now also be written.  Users requested this feature, even though Picard's `SamToFastq` can be used on the outputs of this tool.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -312,12 +363,37 @@ val qualityFormat: Option[FastqQualityFormat] = None,
     sampleInfos.foreach(_.writer.close())
 
+    // Optionally create the FASTQs
+    if (outputFastqs) {
```
Ok, this feels _really_ weird to me.  I figured the idea here was to add an option to output _either_ BAMs or FASTQs.  I find it hard to believe that anyone would really want both - i.e. if I prefer to demux to fastq files, then I'm probably going to ignore the BAM files and vice versa.  Shouldn't the options be exclusive?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -312,12 +363,37 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+    // Optionally create the FASTQs
+    if (outputFastqs) {
+      logger.info("Writing FASTQs")
+      val fastqProgress = new ProgressLogger(this.logger, unit=1e6.toInt)
```
If things stay this way you probably want to give different noun/verbs here so that it's clear what the progress is for.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -312,12 +363,37 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+        val bamReader      = SamReaderFactory.makeDefault().open(info.outputBam)
+        import com.fulcrumgenomics.FgBioDef.javaIterableToIterator
+        bamReader.foreach { bamRecord =>
+          val fastqRecord = FastqRecord(
```
This appears to write all the reads to a single fastq file, without their read number suffixes.  Shouldn't this either output to a pair of fastq files, or supply `readNumber` below?

tfenne commented on src/main/scala/com/fulcrumgenomics/util/DelimitedDataParser.scala:
```diff
@@ -65,6 +65,12 @@ class Row private[util] (private val headerIndices: Map[String,Int], private val
+
+  /** Gets a value from the specified column. If the no column with the given name exists, or the value is null or empty
+    * returns None. */
+  def getOrNone[A](column: String)(implicit tag: ru.TypeTag[A]): Option[A] = if (headerIndices.keys.exists(_ == column)) {
```
A few thoughts:
1) This needs a clearer name since `get()` already returns `Option`. How about instead of a separate method, having `allowMissingColumn: Boolean = false` on `get()`?
2) We could also just make this a parameter to delimited data parser, so it's a mode for the parser?
3) I think you could implement the current version as `headerIndices.get(column).map(get[A](_))` and use the fact that map.get returns an `Option`.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -164,11 +168,13 @@ object DemuxFastqs {
+      |of sample id, sample name, and sample barcode bases (expected not observed), delimited by "-".  A metrics file
+      |will also be output providing analogous information to the metric described here:
       |https://broadinstitute.github.io/picard/picard-metric-definitions.html#SampleBarcodeMetric
+      |In addition to BAM files, gzipped FASTQs can be written using the "--output-fastqs=true" option.
```
Now generates either FASTQs or BAM, but not both.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -193,31 +207,43 @@ object DemuxFastqs {
 (@arg(flag="i", doc="One or more input fastq files each corresponding to a sub-read (ex. index read, read one, read two, fragment).") val inputs: Seq[PathToFastq],
  @arg(flag="o", doc="The output directory in which to place sample BAMs.") val output: DirPath,
- @arg(flag="s", doc="The sample sheet.") val sampleSheet: FilePath,
+ @arg(flag="s", doc="The sample sheet if the metadata CSV file is not given.") val sampleSheet: Option[FilePath] = None,
```
I thought of parsing it both ways but I want to inform the user if either file is malformed along with a useful error message (for the sample sheet at least).

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -238,31 +264,56 @@ val qualityFormat: Option[FastqQualityFormat] = None,
-  val pairedEnd: Boolean = readStructures.count(_.template.nonEmpty) match {
-    case 1 => false
-    case 2 => true
+  readStructures.count(_.template.nonEmpty) match {
```
readStructures is a `Seq[ReadStructure]`, while `_.template` returns a list of all template segments (`Seq[ReadSegment]`) in that read structure, so we count how many read structures have at least one read segment that is a template.  I think it handles the case you're think about.

nh13 commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -238,31 +264,56 @@ val qualityFormat: Option[FastqQualityFormat] = None,
   Io.assertReadable(sampleSheet)
   Io.assertWritableDirectory(output)
 
+  /** Create a sample sheet from either the input sample sheet path or the metadata CSV. */
```
💥

nh13 commented:
@tfenne I think I have addressed all your comments.  By changing the behavior to be either BAM or FASTQ but not both, I was able to some more refactoring, hence the number of changes.

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -44,24 +46,28 @@ import htsjdk.samtools.{SAMRecord, _}
 
   /** Creates the sample output BAM path for the given sample. */
-  private[fastq] def sampleOutputBam(output: DirPath, sample: Sample): PathToBam = {
+  private[fastq] def outputPrefixFrom(output: DirPath, sample: Sample): PathToBam = {
```
I know they're only typedefs, but could the return type get changed to `PathPrefix` as well please?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -238,33 +264,52 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+  private val solexaQualityConverter: SolexaQualityConverter = SolexaQualityConverter.getSingleton
+
+  /** Create a sample sheet from either the input sample sheet path or the metadata CSV. */
+  private val sampleSheet: SampleSheet = {
```
So nice!

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -238,33 +264,52 @@ val qualityFormat: Option[FastqQualityFormat] = None,
     val samples                = samplesFromSampleSheet.toSeq :+ unmatchedSample(samplesFromSampleSheet.size, this.readStructures)
-    val sampleInfos            = samples.map(toSampleInfo)
+    val sampleInfos            = samples.map(sample => SampleInfo(sample, sample.sampleName == UnmatchedSampleId))
+    val sampleToWriter         = sampleInfos.map { info =>  (info.sample, toWriter(info)) }.toMap
```
The two are equivelant, but I kinda prefer this syntax for map creation, how about you:
`sampleInfos.map { info =>  info.sample -> toWriter(info) }.toMap`

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -302,60 +346,138 @@ val qualityFormat: Option[FastqQualityFormat] = None,
-        progress.record(rec)
+
+        val readQualsBytes = rec.quals.getBytes
+        convertQuality(readQualsBytes, qualityFormat, rec.name)
```
Maybe we should note in the usage that even if you are outputting fastq, your quality scores will be standardized to sanger/SAM format?

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -302,60 +346,138 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+      record.setMateUnmappedFlag(true)
+      if (rec.readNumber == 1) record.setFirstOfPairFlag(true) else record.setSecondOfPairFlag(true)
+    }
+    record.setAttribute("BC", rec.sampleBarcode.orNull)
```
For this and the molecularBarcode below why not use the usual?:
`rec.sampleBarcode.foreach(bc => record.setAttribute("BC", bc))`

tfenne commented on src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala:
```diff
@@ -302,60 +346,138 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+/** A writer that writes [[DemuxRecord]]s as [[FastqRecord]]s. */
+private class FastqRecordWriter(prefix: PathPrefix, val pairedEnd: Boolean) extends DemuxWriter {
+  private val writers: IndexedSeq[FastqWriter] = {
+    def f(ext: String): PathToFastq = PathUtil.pathTo(prefix + ext)
```
I think this could be a little cleaner as:
```scala
  (if (pairedEnd) Seq("_R1.fastq.gz", "_R2.fastq.gz") else Seq(".fastq.gz"))
    .map(s => FastqWriter(Io.toWriter(PathUtil.pathTo(prefix + ext))))
    .toIndexedSeq
```

At the very least it could use our `Io.toWriter` instead of `IOUtil.openFileForBufferedWriting`

nh13 replied:
Actually, `IOUtil. openFileForBufferedWriting` supports gzip output, while `Io.toWriter` does not, so I can add `Io.toWriter` to fgbio's `Io`.

nh13 replied:
My bad, it's overriden in `toOutputStream`
